File: src/application/useCases/setEconomicKind.ts
import { suggestEconomicKind, isBulkConfirmable, type KindSuggestion } from '../../domain/suggestEconomicKind'
import { isConsistentWithObservedDirection, type EconomicKind } from '../../domain/entities/economicKind'
import type { Id, Transaction } from '../../domain/entities/types'
import type {
  CategoryRepository,
  Clock,
  TransactionRepository,
  UnitOfWork,
} from '../ports/repositories'

/**
 * SetEconomicKind — تحديد النوع الاقتصادي للعمليات.
 *
 * spec/02: «التصنيف الآلي يقترح، و**تأكيد المستخدم** يحمي اختياره من
 *           الكتابة الآلية اللاحقة.»
 *
 * لذلك لا توجد هنا دالة «طبّق الاقتراحات تلقائيًا». الاقتراح يُعرض،
 * والمستخدم يوافق — إما فرديًا أو جماعيًا على مجموعة يراها.
 */

export interface SuggestionLine {
  transaction: Transaction
  suggestion: KindSuggestion
}

export interface SuggestionSummary {
  /** اقتراحات قاطعة قابلة للتأكيد الجماعي. */
  confirmable: SuggestionLine[]
  /** اقتراحات مرجّحة تحتاج نظرة قبل التأكيد. */
  needsLook: SuggestionLine[]
  /** غامضة: بلا اقتراح، قرار فردي مطلوب. */
  ambiguous: SuggestionLine[]
  /** محدَّدة بالفعل ولا تُمس. */
  alreadySet: number
}

export interface SetEconomicKindDeps {
  txns: TransactionRepository
  /** لترجمة categoryId إلى اسم يفهمه محرك الاقتراح. */
  categories: CategoryRepository
  uow: UnitOfWork
  clock: Clock
}

export class EconomicKindError extends Error {}

export function makeSetEconomicKind(deps: SetEconomicKindDeps) {
  /** يبني الاقتراحات مجمّعة حسب قوتها. **لا يكتب شيئًا.** */
  async function summarize(transactions: readonly Transaction[]): Promise<SuggestionSummary> {
    const categoryNameById = new Map((await deps.categories.listAll()).map((c) => [c.id, c.name]))

    const confirmable: SuggestionLine[] = []
    const needsLook: SuggestionLine[] = []
    const ambiguous: SuggestionLine[] = []
    let alreadySet = 0

    for (const transaction of transactions) {
      if (transaction.economicKindConfirmed) {
        alreadySet++
        continue
      }

      const suggestion = suggestEconomicKind({
        direction: transaction.observedDirection,
        ...(transaction.sourceCategory !== undefined
          ? { sourceCategory: transaction.sourceCategory }
          : {}),
        // ما يعرفه التطبيق عن التصنيف دليل أيضًا، لا يُهدر
        ...(transaction.categoryId && categoryNameById.has(transaction.categoryId)
          ? { categoryName: categoryNameById.get(transaction.categoryId)! }
          : {}),
        ...(transaction.sourceOperationType !== undefined
          ? { sourceOperationType: transaction.sourceOperationType }
          : {}),
        ...(transaction.rawMerchantName !== undefined
          ? { merchantName: transaction.rawMerchantName }
          : {}),
        ...(transaction.rawDescription !== undefined
          ? { description: transaction.rawDescription }
          : {}),
      })

      const line = { transaction, suggestion }
      if (isBulkConfirmable(suggestion)) confirmable.push(line)
      else if (suggestion.kind !== null) needsLook.push(line)
      else ambiguous.push(line)
    }

    return { confirmable, needsLook, ambiguous, alreadySet }
  }

  /**
   * يحدّد نوع عملية واحدة بقرار صريح من المستخدم.
   *
   * يتحقق من اتساق النوع مع اتجاه السيولة الملاحظ، ويرفض التناقض الصريح:
   * لا يجوز وسم حركة صادرة كـ«مرتب» مثلًا.
   */
  async function setOne(transactionId: Id, kind: EconomicKind): Promise<void> {
    const [transaction] = await deps.txns.findByIds([transactionId])
    if (!transaction) throw new EconomicKindError(`عملية غير موجودة: ${transactionId}`)

    if (!isConsistentWithObservedDirection(kind, transaction.observedDirection)) {
      const observed = transaction.observedDirection === 'in' ? 'وارد' : 'صادر'
      throw new EconomicKindError(
        `النوع ده ما يتوافقش مع اتجاه الحركة. الكشف بيقول العملية دي «${observed}».`,
      )
    }

    await deps.txns.update(transactionId, {
      economicKind: kind,
      economicKindConfirmed: true,
      updatedAt: deps.clock.nowIso(),
    })
  }

  /**
   * تأكيد جماعي لاقتراحات قاطعة **اختارها المستخدم صراحةً**.
   *
   * `lineIds` تأتي من الشاشة بعد أن يرى المستخدم القائمة ويوافق.
   * لا تُستدعى تلقائيًا في أي مكان.
   */
  async function confirmBulk(
    transactions: readonly Transaction[],
    selectedIds: readonly Id[],
  ): Promise<{ applied: number; skipped: number }> {
    const wanted = new Set(selectedIds)
    const summary = await summarize(transactions)

    // التأكيد الجماعي مقصور على القاطع — الغامض لا يُؤكَّد جماعيًا أبدًا
    const eligible = summary.confirmable.filter((l) => wanted.has(l.transaction.id))
    const skipped = wanted.size - eligible.length

    if (eligible.length === 0) return { applied: 0, skipped }

    const now = deps.clock.nowIso()
    return deps.uow.run(async () => {
      for (const line of eligible) {
        await deps.txns.update(line.transaction.id, {
          economicKind: line.suggestion.kind!,
          economicKindConfirmed: true,
          updatedAt: now,
        })
      }
      return { applied: eligible.length, skipped }
    })
  }

  return { summarize, setOne, confirmBulk }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/SetEconomicKind.kt
package app.masroufy.usecase

import app.masroufy.core.Direction
import app.masroufy.core.EconomicKind
import app.masroufy.core.Id
import app.masroufy.core.KindSuggestion
import app.masroufy.core.SuggestionInput
import app.masroufy.core.Transaction
import app.masroufy.core.isBulkConfirmable
import app.masroufy.core.isConsistentWithObservedDirection
import app.masroufy.core.suggestEconomicKind
import app.masroufy.port.CategoryRepository
import app.masroufy.port.Clock
import app.masroufy.port.TransactionPatch
import app.masroufy.port.TransactionRepository
import app.masroufy.port.UnitOfWork

/**
 * SetEconomicKind — نقل `setEconomicKind.ts`. spec/02: «التصنيف الآلي يقترح،
 * و**تأكيد المستخدم** يحمي اختياره». مفيش هنا «طبّق الاقتراحات تلقائيًا» —
 * الاقتراح بيتعرض والمستخدم بيوافق، فرديًا أو جماعيًا على اللي شايفه.
 */

data class SuggestionLine(val transaction: Transaction, val suggestion: KindSuggestion)

data class SuggestionSummary(
    /** اقتراحات قاطعة قابلة للتأكيد الجماعي. */
    val confirmable: List<SuggestionLine>,
    /** اقتراحات مرجّحة محتاجة نظرة قبل التأكيد. */
    val needsLook: List<SuggestionLine>,
    /** غامضة: من غير اقتراح، قرار فردي مطلوب. */
    val ambiguous: List<SuggestionLine>,
    /** محدَّدة بالفعل وما بتتمسش. */
    val alreadySet: Int,
)

data class BulkConfirmResult(val applied: Int, val skipped: Int)

data class SetEconomicKindDeps(
    val txns: TransactionRepository,
    /** لترجمة categoryId لاسم يفهمه محرك الاقتراح. */
    val categories: CategoryRepository,
    val uow: UnitOfWork,
    val clock: Clock,
)

class SetEconomicKind(private val deps: SetEconomicKindDeps) {
    /** بيبني الاقتراحات متجمّعة حسب قوتها. **ما بيكتبش حاجة.** */
    suspend fun summarize(transactions: List<Transaction>): SuggestionSummary {
        val categoryNameById = deps.categories.listAll().associate { it.id to it.name }

        val confirmable = mutableListOf<SuggestionLine>()
        val needsLook = mutableListOf<SuggestionLine>()
        val ambiguous = mutableListOf<SuggestionLine>()
        var alreadySet = 0

        for (transaction in transactions) {
            if (transaction.economicKindConfirmed) {
                alreadySet++
                continue
            }

            val suggestion = suggestEconomicKind(
                SuggestionInput(
                    direction = transaction.observedDirection,
                    sourceCategory = transaction.sourceCategory,
                    // اللي التطبيق عارفه عن التصنيف دليل كمان، ما يتهدرش
                    categoryName = transaction.categoryId?.let { categoryNameById[it] },
                    sourceOperationType = transaction.sourceOperationType,
                    merchantName = transaction.rawMerchantName,
                    description = transaction.rawDescription,
                ),
            )

            val line = SuggestionLine(transaction, suggestion)
            when {
                isBulkConfirmable(suggestion) -> confirmable += line
                suggestion.kind != null -> needsLook += line
                else -> ambiguous += line
            }
        }

        return SuggestionSummary(confirmable, needsLook, ambiguous, alreadySet)
    }

    /**
     * بيحدد نوع عملية واحدة بقرار صريح من المستخدم، وبيرفض التناقض الصريح
     * مع اتجاه السيولة الملاحظ: مينفعش حركة صادرة تتعلم «مرتب».
     */
    suspend fun setOne(transactionId: Id, kind: EconomicKind) {
        val transaction = deps.txns.findByIds(listOf(transactionId)).firstOrNull()
            ?: throw IllegalStateException("عملية غير موجودة: $transactionId")

        if (!isConsistentWithObservedDirection(kind, transaction.observedDirection)) {
            val observed = if (transaction.observedDirection == Direction.IN) "وارد" else "صادر"
            throw IllegalStateException("النوع ده ما يتوافقش مع اتجاه الحركة. الكشف بيقول العملية دي «$observed».")
        }

        deps.txns.update(
            transactionId,
            TransactionPatch(economicKind = kind, economicKindConfirmed = true, updatedAt = deps.clock.nowIso()),
        )
    }

    /**
     * تأكيد جماعي لاقتراحات قاطعة **اختارها المستخدم صراحةً** —
     * ما بيتندهش تلقائيًا في أي مكان، والغامض عمره ما بيتأكد جماعيًا.
     */
    suspend fun confirmBulk(transactions: List<Transaction>, selectedIds: List<Id>): BulkConfirmResult {
        val wanted = selectedIds.toSet()
        val summary = summarize(transactions)

        val eligible = summary.confirmable.filter { it.transaction.id in wanted }
        val skipped = wanted.size - eligible.size

        if (eligible.isEmpty()) return BulkConfirmResult(applied = 0, skipped = skipped)

        val now = deps.clock.nowIso()
        return deps.uow.run {
            for (line in eligible) {
                deps.txns.update(
                    line.transaction.id,
                    TransactionPatch(economicKind = line.suggestion.kind!!, economicKindConfirmed = true, updatedAt = now),
                )
            }
            BulkConfirmResult(applied = eligible.size, skipped = skipped)
        }
    }
}
